### PR TITLE
esp, rx, add support for esp32 receivers

### PR DESCRIPTION
**Pull request by tmcadam**
WORK IN PROGRESS : DO NOT MERGE

This PR is to add support for ESP32 based True Diversity ELRS receivers.

### CODE REVIEW

**Comment by olliw42** [comment]
why does this want to go into main? 
is dev-esp-rx-td finsihed and sorted out already?

**Comment by olliw42** [comment]
sorry, I'm still a bit puzzled about what we want to do here
in my naive view, in order to advance dev-esp-rx-td branch, we PR to the dev-esp-rx-td branch. Only once the dev-esp-rx-td branch is in a state to be merged to main, or has added commits which shall be brought over to main, one would PR it to bring to main. The discussion around dev-esp-rx-td branch ergo wouldn't be in such a PR since ... kind of by definition dev-esp-rx-td branch was considered ok for merge.
like so, isn't it a duplication?

**Comment by tmcadam** [comment]
> sorry, I'm still a bit puzzled about what we want to do here in my naive view, in order to advance dev-esp-rx-td branch, we PR to the dev-esp-rx-td branch. Only once the dev-esp-rx-td branch is in a state to be merged to main, or has added commits which shall be brought over to main, one would PR it to bring to main. The discussion around dev-esp-rx-td branch ergo wouldn't be in such a PR since ... kind of by definition dev-esp-rx-td branch was considered ok for merge. like so, isn't it a duplication?
> 
> I would think we want things in dev-esp-rx-td for as long as it is under dev and/or testing. Not sure what your idea is.

Yes, you understand correctly. All development will take place in `dev-esp-rx-td`. We all contribute changes to the branch `dev-esp-rx-td` through PRs from our own branches. 

It's not a duplication, it's another layer of PR. This PR is just automatically tracking the changes `dev-esp-rx-td` i.e. everyone's contributions in one place. They can be reviewed together here, this PR is tracking the readiness of the `dev-esp-rx-td` to merge to `main`.

When `dev-esp-rx-td` is "ready", maybe some days or weeks away. We can approve and merge here.

**Comment by olliw42** [comment]
I'm sorry, but I still don't get it. To me it still looks like an unneeded and confusing duplication with no benefit. But I will let you guys do your thing and observe what this is going to be, I maybe then understand :)

**Comment by tmcadam** [comment]
> I'm sorry, but I still don't get it. To me it still looks like an unneeded and confusing duplication with no benefit. But I will let you guys do your thing and observe what this is going to be, I maybe then understand :)

Thanks, lets try it and see, if it isn't helping matters we use another approach. Apologies for my poor explanation!

File: mLRS/Common/leds.h [hunk 1]
```diff
@@ -39,7 +39,9 @@ class tLEDs
 #ifndef DEVICE_HAS_SINGLE_LED_RGB
             if (!blink) { led_green_toggle(); led_red_toggle(); }
 #endif
+#ifdef DEVICE_HAS_SINGLE_LED_RGB
```
Comment (olliw42) [inline mLRS/Common/leds.h:42]: I'm not sure yet if I should like the introduction of yet another color, but could you in Tick_ms() separate the code into the three sections
#ifdef DEVICE_HAS_SINGLE_LED
....
#elif defined DEVICE_HAS_SINGLE_LED_RGB
....
#esle
...
#endif 
it's getting complicated to follow the flow of the code

my hesitation with yet another blink pattern is that I don't like having that many different ones ... noone can memorize them all ... example, could you tell all ELRS rgb blink patterns without having to look em up?
is there a specific reason to not go with red-green blink?

Comment (jlpoltrack) [reply mLRS/Common/leds.h]: More of a fun one, no plans for any more blink patterns :).   

At least in English the alliteration makes sense to me, **_b_** lue for _**b**_ ind

**Comment by jlpoltrack** [comment]
FLRC stress test (2x recommended stream rates):

image

**Comment by jlpoltrack** [comment]
Docs branch: https://github.com/olliw42/mLRS-docu/tree/esp32-true-diversity

File: mLRS/Common/hal/esp/rx-hal-generic-2400-td-pa-esp32.h [hunk 2]
```diff
@@ -106,13 +106,24 @@ IRQHANDLER(void SX2_DIO_EXTI_IRQHandler(void);)
...
 {
+    gpio_init(SX2_CS_IO, IO_MODE_OUTPUT_PP_HIGH);
     gpio_init(SX2_DIO1, IO_MODE_INPUT_ANALOG);
     gpio_init(SX2_BUSY, IO_MODE_INPUT_PU);
     gpio_init(SX2_TX_EN, IO_MODE_OUTPUT_PP_LOW);
     gpio_init(SX2_RX_EN, IO_MODE_OUTPUT_PP_LOW);
     gpio_init(SX2_RESET, IO_MODE_OUTPUT_PP_LOW);
 }
 
+IRAM_ATTR static inline void sx2_select(void)
```
Comment (olliw42) [inline mLRS/Common/hal/esp/rx-hal-generic-2400-td-pa-esp32.h:117]: I like that better, moving it to the hal
would you mind calling it spi2_select/deselct() nevertheless
should make the sx driver canonical

File: mLRS/Common/hal/esp-glue.h [hunk 3]
```diff
@@ -18,16 +18,32 @@
...
+{
 #ifdef ESP32
     taskENTER_CRITICAL(&esp32_spinlock);
 #elif defined ESP8266
     noInterrupts();
 #endif
 }
 
-void __enable_irq(void)
+void exitCritical(void)
```
Comment (olliw42) [inline mLRS/Common/hal/esp-glue.h:46]: that's in principle not a bad idea, but I would suggest that we just don't try for now to do anything smart. That is, that we don't define any such functions, keep __enable_irq()/__disable_irq() empty, and instead code what is needed at the places where it is needed
will not look super nice for now, but at least allows us to do what it is needed where something is needed exactly as it is needed
the  __enable_irq()/__disable_irq() can also interfer with tx stm32 code, so let's keep them empty

Comment (jlpoltrack) [reply mLRS/Common/hal/esp-glue.h]: Should be addressed now

File: mLRS/Common/hal/hal.h [hunk 4]
```diff
@@ -243,7 +243,6 @@ Note: Some "high-level" features are set for each device in the device_conf.h fi
 
 #ifdef RX_ELRS_GENERIC_2400_TD_PA_ESP32
 #include "esp/rx-hal-generic-2400-td-pa-esp32.h"
-#endif
```
Comment (olliw42) [inline mLRS/Common/hal/hal.h]: this does not look right :)

**Comment by jlpoltrack** [comment]
Tested on bench with latest commit on Matek ELRS-D and BetaFPV Super - looks good.

**Comment by tmcadam** [comment]
Bench tested with R9M and Generic 900 RX. no issues seen. This is looking really good now. Amazing effort @jlpoltrack and @olliw42 

```
100(100),100, -037,-024,  004; 00000, 00653;
100(100),100, -036,-025,  003; 00000, 00653;
100(100),100, -037,-024,  004; 00000, 00693;
100(100),100, -063,-054,  004; 00000, 00693;
100(100),100, -036,-025,  004; 00000, 00986;
100(100),100, -037,-024,  003; 00000, 00986;
100(100),100, -037,-025,  003; 00000, 00653;
100(100),100, -036,-024,  004; 00000, 00653;
100(100),100, -067,-051,  003; 00000, 00653;
100(100),100, -036,-023,  004; 00000, 00653;
100(100),100, -036,-023,  003; 00000, 00655;
100(100),100, -034,-022,  004; 00000, 00653;
100(100),100, -036,-024,  003; 00000, 00653;
100(100),100, -053,-042,  003; 00000, 00653;
100(100),100, -035,-023,  003; 00000, 00653;
100(100),100, -035,-023,  004; 00000, 00653;
100(100),100, -036,-024,  004; 00000, 00653;
100(100),100, -035,-023,  002; 00000, 00679;
100(100),100, -052,-040,  003; 00000, 00679;
100(100),100, -036,-023,  004; 00000, 00653;
100(100),100, -036,-023,  003; 00000, 00653;
100(100),100, -035,-023,  004; 00000, 00653;
100(100),100, -036,-023,  003; 00000, 00653;
100(100),100, -052,-040,  003; 00000, 00653;
100(100),100, -035,-023,  003; 00000, 00653;
100(100),100, -036,-023,  003; 00000, 00693;
```

image

**Comment by olliw42** [comment]
I haven't done anything essential here ;)
I will retest with my RP4TD this evening, and then merge ...

**Comment by olliw42** [comment]
tested on bench for
- RM RP4TD
- betafpv nano 2.4 rx
- betafpv nano 900 rx

will wait few hours, and if no complains will merge :):):)

**Comment by olliw42** [comment]
Co-authored-by: jlpoltrack <41841496+jlpoltrack@users.noreply.github.com>

**Pull request merged**